fvm/fvm.go

```diff
@@ -26,17 +27,36 @@ func New(rt runtime.Runtime) *VirtualMachine {
 }
 
 // Run runs a procedure against a ledger in the given context.
-func (vm *VirtualMachine) Run(ctx Context, proc Procedure, ledger state.Ledger) error {
+func (vm *VirtualMachine) Run(ctx Context, proc Procedure, ledger state.Ledger) (err error) {
 
 	st := state.NewState(ledger,
 		state.WithMaxKeySizeAllowed(ctx.MaxStateKeySize),
 		state.WithMaxValueSizeAllowed(ctx.MaxStateValueSize),
 		state.WithMaxInteractionSizeAllowed(ctx.MaxStateInteractionSize))
 
-	err := proc.Run(vm, ctx, st)
+	defer func() {
+		if r := recover(); r != nil {
+
+			// Cadence may fail to encode certain values.
+			// Return an error for now, which will cause transactions to revert.
+			//
+			if encodingErr, ok := r.(interpreter.EncodingUnsupportedValueError); ok {
+				err = &EncodingUnsupportedValueError{
+					Path:  encodingErr.Path,
+					Value: encodingErr.Value,
+				}
+				return
+			}
+
+			panic(r)
+		}
+	}()
```

Comment:
shouldn't this logic be part of https://github.com/onflow/flow-go/blob/master/fvm/errors.go#L280-L300

Reply:
My understanding is that that function handles the case where Cadence called into the FVM (e.g. to get block info), and the FVM side paniced. 

Here, Cadence itself panics with an implementation error that should never occur, but we need to gracefully handle right now.

Is this correct @m4ksio @psiemens ?

Reply:
Yeah, i believe the function that Ramtin linked is specifically to handle any runtime error, that is returned as such, but it doesn't handle panics. We will definitely need an extra recover somewhere, either here in the FVM or in cadence runtime. I'm okay with it being here

Reply:
Yes, I think this is the correct place for this right now.

`handleRuntimeError` is only called on errors that are explicitly returned by `vm.Runtime.ExecuteTransaction`. The `handleRuntimeError` function call is made through `handleError`, which is called here: https://github.com/onflow/flow-go/blob/f0b8b92ebce1f48f0d64e8e4243893834478bbcd/fvm/transaction.go#L43

As @turbolent pointed out, in this case we're dealing with an unexpected panic from `vm.Runtime.ExecuteTransaction` that will bubble up before reaching the `handleError` invocation.

I think `Run()` is the best place for this recovery because it will work for both scripts and transactions. Ideally this is temporary until we refactor the error handling logic.

Reply:
I see